pkg/build/util/util.go

```diff
@@ -167,3 +171,195 @@ func VersionForBuild(build *buildapi.Build) int {
+	namespace := defaultNamespace
+	if len(from.Namespace) > 0 {
+		namespace = from.Namespace
+	}
+
+	glog.V(4).Infof("Resolving ImageStreamReference %s of Kind %s in namespace %s", from.Name, from.Kind, namespace)
+	switch from.Kind {
+	case "ImageStreamImage":
+		imageStreamImage, err := client.GetImageStreamImage(kapi.WithNamespace(ctx, namespace), from.Name)
+		if err != nil {
+			glog.V(2).Infof("Error ImageStreamReference %s in namespace %s: %v", from.Name, namespace, err)
+			if errors.IsNotFound(err) {
+				return "", err
+			}
+			return "", err
+		}
+		image := imageStreamImage.Image
+		glog.V(4).Infof("Resolved ImageStreamReference %s to image %s with reference %s in namespace %s", from.Name, image.Name, image.DockerImageReference, namespace)
+		return image.DockerImageReference, nil
+	case "ImageStreamTag":
+		name := strings.Split(from.Name, ":")[0]
+		is, err := client.GetImageStream(kapi.WithNamespace(ctx, namespace), name)
+		if err != nil {
+			glog.V(2).Infof("Error getting ImageStream %s/%s: %v", namespace, name, err)
+			if errors.IsNotFound(err) {
+				return "", err
+			}
+			return "", err
+		}
+		image, err := imageapi.DockerImageReferenceForStream(is)
+		if err != nil {
+			glog.V(2).Infof("Error resolving Docker image reference for %s/%s: %v", namespace, name, err)
+			return "", err
+		}
+		glog.V(4).Infof("Resolved ImageStreamTag %s/%s to repository %s", namespace, from.Name, image)
+		return image.String(), nil
+	case "DockerImage":
+		return from.Name, nil
+	default:
+		return "", fmt.Errorf("Unknown From Kind %s", from.Kind)
+	}
+}
+
+// ResolveImageStreamReference looks up the ImageStream[Tag/Image] and converts it to a
+// docker pull spec that can be used in an Image field.
+func ResolveImageStreamReference(client BuildClient, ctx kapi.Context, from kapi.ObjectReference, defaultNamespace string) (string, error) {
+	var namespace string
+	if len(from.Namespace) != 0 {
+		namespace = from.Namespace
+	} else {
+		namespace = defaultNamespace
+	}
+
+	glog.V(4).Infof("Resolving ImageStreamReference %s of Kind %s in namespace %s", from.Name, from.Kind, namespace)
+	switch from.Kind {
+	case "ImageStreamImage":
+		imageStreamImage, err := client.GetImageStreamImage(kapi.WithNamespace(ctx, namespace), from.Name)
+		if err != nil {
+			glog.V(2).Infof("Error ImageStreamReference %s in namespace %s: %v", from.Name, namespace, err)
+			if errors.IsNotFound(err) {
+				return "", err
+			}
+			return "", err
+		}
+		image := imageStreamImage.Image
+		glog.V(4).Infof("Resolved ImageStreamReference %s to image %s with reference %s in namespace %s", from.Name, image.Name, image.DockerImageReference, namespace)
+		return image.DockerImageReference, nil
+	case "ImageStreamTag":
+		imageStreamTag, err := client.GetImageStreamTag(kapi.WithNamespace(ctx, namespace), from.Name)
+		if err != nil {
+			glog.V(2).Infof("Error resolving ImageStreamTag reference %s in namespace %s: %v", from.Name, namespace, err)
```

Comment:
will move it below the existing isNotFound return, but just a heads up, I really only opened this PR to get rosie to do some testing, it's still very much a WIP.